src/DotVVM.Framework/ViewModel/Serialization/DotvvmDictionaryConverter.cs

```diff
@@ -0,0 +1,80 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using DotVVM.Framework.Utils;
+using Newtonsoft.Json;
+
+namespace DotVVM.Framework.ViewModel.Serialization
+{
+    public class DotvvmDictionaryConverter : JsonConverter
+    {
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            if (value == null)
+            {
+                writer.WriteNull();
+            }
+            else
+            {
+                var dict = value as IDictionary;
+
+                var attrs = value.GetType().GetGenericArguments();
+                var keyValuePair = typeof(KeyValuePair<,>).MakeGenericType(attrs);
+                var listType = typeof(List<>).MakeGenericType(keyValuePair);
+
+
+                var keyEnumerator = dict.Keys.GetEnumerator();
+                var valuesEnumerator = dict.Values.GetEnumerator();
+
+
+                var list = Activator.CreateInstance(listType);
+                var invokeMethod = listType.GetMethod(nameof(List<object>.Add));
+                while (keyEnumerator.MoveNext() && valuesEnumerator.MoveNext())
+                {
+                    var item = Activator.CreateInstance(keyValuePair, keyEnumerator.Current, valuesEnumerator.Current);
+                    invokeMethod.Invoke(list, new[] { item });
+                }
+
+                serializer.Serialize(writer, list);
+            }
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            if (reader.TokenType == JsonToken.Null)
+            {
+                return null;
+            }
+            else if (reader.TokenType == JsonToken.StartArray)
+            {
+
+                var attrs = objectType.GetGenericArguments();
+                var keyValuePair = typeof(KeyValuePair<,>).MakeGenericType(attrs);
+                var listType = typeof(List<>).MakeGenericType(keyValuePair);
+
+                var dict = existingValue as IDictionary;
+                dict ??= (IDictionary)Activator.CreateInstance(objectType);
```

Comment:
This would be problematic if the `objectType` is `IDictionary<...>`.

Reply:
I have added condition `typeof(IDictionary).IsAssignableFrom(objectType)` to `bool CanConvert(Type objectType)` that prevents (de)serialization of instance that implements only `IDictionary<,>`. I am not even sure whether that is something we want to support until standard Dictionary<,> is fully supported.